source/grid/manifold.cc

```diff
@@ -62,6 +62,128 @@ get_new_point (const Quadrature<spacedim> &quad) const
 }
 
 
+
+template <>
+Tensor<1,2>
+Manifold<2, 2>::
+normal_vector (const typename Triangulation<2, 2>::face_iterator &face,
+               const Point<2> &p) const
+{
+  const int spacedim=2, dim=2;
+
+  Tensor<1,spacedim> tangent = ((p-face->vertex(0)).norm_square() > (p-face->vertex(1)).norm_square() ?
+                                get_tangent_vector(p, face->vertex(0)) :
+                                -get_tangent_vector(p, face->vertex(1)));
+  return cross_product_2d(tangent);
+}
+
+template<>
+Tensor<1,3>
+Manifold<3, 3>::
+normal_vector (const typename Triangulation<3, 3>::face_iterator &face,
+               const Point<3> &p) const
+{
+  const int spacedim=3, dim=3;
+  Tensor<1,spacedim> t1,t2;
+
+  // Take the difference between p and all four vertices
+  Tensor<1,spacedim> dp[4];
+  double dpns[4];
+  int min_index=-1;
+  double min_distance = 0;
+  for (unsigned int i=0; i<4; ++i)
+    {
+      dp[i] = p-face->vertex(i);
+      dpns[i] = dp[i].norm_square();
+      min_index = (min_index == -1 ? (int)i : dpns[i] < min_distance ? i : min_index);
```

Comment:
you can avoid the outer `?:` operator if you initialize `min_index` with 0, initialize `min_distance` with the distance to vertex 0, and then only run the loop from 1 to 4, instead of 0 to 4.